admission_control/admission-control-service/src/upstream_proxy.rs

```diff
@@ -6,198 +6,238 @@ use admission_control_proto::proto::admission_control::{
+                    .await;
+            },
+            network_event = events.select_next_some() => {
+                match network_event {
+                    Ok(event) => {
+                        match event {
+                            Event::NewPeer(peer_id) => {
+                                debug!("[admission control] new peer {}", peer_id);
+                                new_peer(&mut peer_info, peer_id);
+                            }
+                            Event::LostPeer(peer_id) => {
+                                debug!("[admission control] lost peer {}", peer_id);
+                                lost_peer(&mut peer_info, peer_id);
+                            }
+                            Event::RpcRequest((peer_id, mut message, callback)) => {
+                                if let Some(AdmissionControlMsg_oneof::SubmitTransactionRequest(request)) = message.message {
+                                    let peer_id = pick_peer(peer_info.clone());
+                                    bounded_executor
+                                        .spawn(process_submit_transaction_request(upstream_proxy_data.clone(), peer_id, request, callback))
+                                        .await;
                                 }
-                                _ => {},
                             }
-                        },
-                        Err(err) => { error!("[admission control] network error {:?}", err); },
-                    }
+                            _ => {},
+                        }
+                    },
+                    Err(err) => { error!("[admission control] network error {:?}", err); },
                 }
             }
         }
     }
+}
 
-    /// new peer discovery handler
-    /// adds new entry to `peer_info`
-    fn new_peer(&mut self, peer_id: PeerId) {
-        if let Some(state) = self.peer_info.get_mut(&peer_id) {
-            *state = true;
-        }
+/// new peer discovery handler
+/// adds new entry to `peer_info`
+fn new_peer(peer_info: &mut HashMap<PeerId, bool>, peer_id: PeerId) {
+    if let Some(state) = peer_info.get_mut(&peer_id) {
+        *state = true;
     }
+}
 
-    /// lost peer handler. Marks connection as dead
-    fn lost_peer(&mut self, peer_id: PeerId) {
-        if let Some(state) = self.peer_info.get_mut(&peer_id) {
-            *state = false;
-        }
+/// lost peer handler. Marks connection as dead
+fn lost_peer(peer_info: &mut HashMap<PeerId, bool>, peer_id: PeerId) {
+    if let Some(state) = peer_info.get_mut(&peer_id) {
+        *state = false;
+    }
+}
+
+fn pick_peer(peer_info: HashMap<PeerId, bool>) -> Option<PeerId> {
+    let active_peer_ids = get_active_upstream_peers(&peer_info);
+    if !active_peer_ids.is_empty() {
+        return active_peer_ids.choose(&mut rand::thread_rng()).cloned();
     }
+    None
+}
+
+fn get_active_upstream_peers(peer_info: &HashMap<PeerId, bool>) -> Vec<PeerId> {
```

Comment:
also can we directly put into `pick_peer` function. Looks like there's only one call site